### ISSUE TITLE
SoundFileView blocks QtCollider while loading file

### ISSUE THREAD

**Issue by miguel-negrao** [supercollider/supercollider#3739]
```
w = SoundFileView();
w.front;

(
x = Process.elapsedTime;
w.load("bigfile.aiff");
(Process.elapsedTime - x).postln;
)
```
With a 3.1GB file the load method only returns after 74 seconds, during this time all other GUI widgets are non responsive and gnome offers to kill QtCollider. Only after load returns does the SoundFileView window start to draw.

I would like there to be a version of load that would return immeditely and do any file loading and processing in the background without QtCollider ever blocking.

**Comment by muellmusik** [comment]
AIFF files under 2 GB pose no problems for me and return instantly. AIFF files over 2 GB are problematic and not guaranteed to work. Try saving your file as a CAF or WAV64 and see if that makes a difference. A 3.3 GB CAF returns after only a slight delay for me.

**Comment by miguel-negrao** [comment]
slow rotating disk:
2.4GB CAF file -> 94 seconds
1.4GB AIFF file -> 31 seconds

fast ssd:
2.4GB CAF file -> 6.3 seconds
I suppose the delay is caused from reading the file to memory, and it gets worse the slower the disk. Perhaps you have a fast ssd ? So I think the issue is real... IMHO it would be nice if the file reading was be done in the background... it seems currently only the drawing is done in the background ?

**Comment by miguel-negrao** [comment]
Opening the debbugger it seems that sclang is blocking for those seconds above on sf_command:

https://github.com/supercollider/supercollider/blob/cca12ff02a774a9ea212e8883551d3565bb24a6f/QtCollider/widgets/soundfileview/view.cpp#L227

which is calling sf_read_double which is calling the read sys call. So probably this is where the file read happens...

**Comment by muellmusik** [comment]
> which is calling sf_read_double which is calling the read sys call. So probably this is where the file read happens...

Ah okay: 
https://github.com/erikd/libsndfile/blob/7ea3f9d8746000cc82c016d0b5d48452bb80e9fe/src/sndfile.c#L974

		case SFC_SET_SCALE_FLOAT_INT_READ :
			old_value = psf->float_int_mult ;

			psf->float_int_mult = (datasize != 0) ? SF_TRUE : SF_FALSE ;
			if (psf->float_int_mult && psf->float_max < 0.0)
				/* Scale to prevent wrap-around distortion. */
				psf->float_max = (32768.0 / 32767.0) * psf_calc_signal_max (psf, SF_FALSE) ;
			return old_value ;

psf_calc_signal_max has to read the file. So I think putting this into the reading thread will solve this.

**Comment by miguel-negrao** [comment]
The sf_command with [SFC_SET_SCALE_FLOAT_INT_READ](http://www.mega-nerd.com/libsndfile/command.html#SFC_SET_SCALE_FLOAT_INT_READ) internally calls [psf_calc_signal_max](https://github.com/erikd/libsndfile/blob/64e5a3f1824a4cbd7f2d8e03e8032ae1a6cb83c3/src/command.c#L276). This function in some cases reads the whole file to find largest sample. I wonder if in those cases we are not reading the whole file twice as I think the file is read again possibly here:

https://github.com/supercollider/supercollider/blob/cca12ff02a774a9ea212e8883551d3565bb24a6f/QtCollider/widgets/soundfileview/filestream.cpp#L39

I think setting SFC_SET_SCALE_FLOAT_INT_READ might not be needed for integer sound files ? In any case that call should probably be made in a different thread.

**Comment by miguel-negrao** [comment]
Right, we got to the same conclusion. :-)

**Comment by muellmusik** [comment]
> Right, we got to the same conclusion. :-)

Okay, so there a couple of options. Simplest, move that into the read thread. Most complicated, try to rework the reading code so that that call is not necessary, i.e. do your own float scaling. This could be more efficient overall. Compromise, IIUC you only need that call when reading float files, so there's no point doing it if you're not. So you could move it into the read thread, and check the sfinfo struct before applying the sfcommand to scale floats.

**Comment by miguel-negrao** [comment]
Actually, looking again at the code I think if the file is a float we read it as float and calculate ourselves the maximum and therefore the sfcommand is not even needed...:

https://github.com/supercollider/supercollider/blob/cca12ff02a774a9ea212e8883551d3565bb24a6f/QtCollider/widgets/soundfileview/filestream.cpp#L51

If that is the case we can safely delete the sfcommand.

**Comment by miguel-negrao** [comment]
> and calculate ourselves the maximum

Actually, I think we don't calculate the maximum but just clip at -1, 1, but still we read a float which doesn't need the sfcommand.

**Comment by muellmusik** [comment]
> Actually, I think we don't calculate the maximum but just clip at -1, 1, but still we read a float which doesn't need the sfcommand.

So if I understand this correctly, what the sfcommand does is normalise when reading a float file as a short or int. It's not totally clear to me what should be the default behaviour for SFView. Certainly float or double files will not clip, so you may want to see the file normalised, especially if it exceeds +/- 1.0. On the other hand if everything is normalised, we lose comparability.

It would be nice to have either behaviour. Normalisation might be less of an issue if it was at least clear that it had happened.

**Comment by miguel-negrao** [comment]
> So if I understand this correctly, what the sfcommand does is normalise when reading a float file as a short or int.

It think it finds the maximum which will be used later when a read as short or int is done. I think that command can be safely removed, which will remove the current freezing of QtCollider.


> It's not totally clear to me what should be the default behaviour for SFView. Certainly float or double files will not clip, so you may want to see the file normalised, especially if it exceeds +/- 1.0. On the other hand if everything is normalised, we lose comparability.
> 
> It would be nice to have either behaviour. Normalisation might be less of an issue if it was at least clear that it had happened.

Yes, would be nice for it to be a flag. Also, should we normalize relative to the whole file or to the region which we are analyzing/viewing ?

**Comment by miguel-negrao** [comment]
@brianlheim great, had a quick a look and it seems @dyfer is adressing most of the issues mentioned here.

**Comment by dyfer** [comment]
@miguel-negrao yes, I removed reading floats into ints in my PR, everything is read in float. Also, there's no normalization: you can scale the view vertically for large/small values if desired (but you'd need to find min/max yourself if you want to automate this).

**Comment by brianlheim** [comment]
IIUC, the only feature request remaining for this PR is to add a non-blocking read method. Is that correct, @miguel-negrao , @muellmusik ? How difficult would you consider it to be to add this?

**Comment by miguel-negrao** [comment]
@brianlheim no, I think this PR already address that. By removing the sf_command line (which the PR does) all the remaining processing is done in a different thread and therefore the method should return immediately not blocking QtCollider.

**Comment by brianlheim** [comment]
Sorry, I meant to say for this issue, not for this PR!

**Comment by brianlheim** [comment]
Then can this issue be closed?

**Comment by brianlheim** [comment]
I'm trying to determine whether that PR will close this ticket or not.

**Comment by miguel-negrao** [comment]
I believe, from looking at the code, that the PR once merged will close this ticket.

**Comment by muellmusik** [comment]
I concur.

**Comment by jamshark70** [comment]
> I'm trying to determine whether that PR will close this ticket or not.

I'll test in an hour or two.

**Comment by jamshark70** [comment]
Confirmed. #3662 definitely changes SoundFileView reading to be asynchronous.

```
(
var file = SoundFile.openRead("/media/dlm/...stuff.../17-18-fall/lork/christmas-in-iceland.flac"),
view, startTime;

if(file.notNil) {
	view = SoundFileView(nil, Rect.aboutPoint(Window.screenBounds.center, 300, 200)).front;
	startTime = Main.elapsedTime;
	view.soundfile_(file).read(doneAction: {
		"doneAction: % seconds later\n".postf(Main.elapsedTime - startTime);
		file.close;
	});
	"read method returned: % seconds later\n".postf(Main.elapsedTime - startTime);
};
)

read method returned: 0.00050171200001614 seconds later
-> read method returned: 0.00050171200001614 seconds later

doneAction: 1.240169211 seconds later
```

Other notes:

- I tried it while running a fast Pbind. No break in the OSC messages going out.

- The progress display in the window is really nice!

**Comment by miguel-negrao** [comment]
Fixed with #3662.
